src/ast/nodes/nodes.h

```diff
@@ -155,13 +155,48 @@ class FunctionCall : public interface::AstNode<FunctionCall> {
   std::vector<Expression> arguments;
 };
 
+class BinaryOperation: interface::AstNode<BinaryOperation> {
+public:
+  enum class Operation: char { NOT_AN_OPERATION='N', ADD='+', SUBSTRACT='-', MULTIPLY='*', DIVIDE='/' };
+
+  static constexpr const char *node_name = "Node_BinaryOperation";
+
+public:
+  BinaryOperation(Operation op, std::unique_ptr<Expression> &&lhs, std::unique_ptr<Expression> &&rhs)
+    : op(op), lhs(std::move(lhs)), rhs(std::move(rhs)) {}
+
+  BinaryOperation(const BinaryOperation &other)
+    : op(other.op), lhs(std::make_unique<Expression>(*other.lhs)), rhs(std::make_unique<Expression>(*other.rhs)) {}
+
+  inline void genAsm_x86_64(codegen::NasmGenerator_x86_64 &generator) const {
+    (void)generator;
+    THROW("BinaryOperation genAsm_x86_64 should not be called");
+  }
+
+  inline void debug(size_t depth) const;
+
+  inline void decorate(scopes::ScopeStack &scopeStack, scopes::Scope &scope);
+
+  inline void loadValueInRegister(codegen::NasmGenerator_x86_64 &generator, scopes::GeneralPurposeRegister targetRegister) const;
+
+private:
+  Operation op;
+  std::unique_ptr<Expression> lhs;
+  std::unique_ptr<Expression> rhs;
```

Comment:
Why `unique_ptr` ?

Reply:
because the parent should own the children, no one else. basically
1) does anyone else that will survive the lifetime of the parent need the node? -> no (injective)
2) does the child still make sense without the parent -> no (surjective)

Reply:
if you need to reference the child at some point, return an Expression&, or an Expression* in case the child is optional for example. It's totally fine to use non-owning references.

I suggest reading the programmer bible https://doc.rust-lang.org/book/ch04-00-understanding-ownership.html if you're interested lol